src/vm/jitinterface.cpp

```diff
@@ -2286,6 +2286,50 @@ BOOL CEEInfo::checkMethodModifier(CORINFO_METHOD_HANDLE hMethod,
 }
 
 /*********************************************************************/
+static unsigned ComputeGCLayout(MethodTable * pMT, BYTE* gcPtrs)
+{
+    STANDARD_VM_CONTRACT;
+
+    unsigned result = 0;
+
+    _ASSERTE(pMT->IsValueType());
+
+    ApproxFieldDescIterator fieldIterator(pMT, ApproxFieldDescIterator::INSTANCE_FIELDS);
+    for (FieldDesc *pFD = fieldIterator.Next(); pFD != NULL; pFD = fieldIterator.Next())
+    {
+        int fieldStartIndex = pFD->GetOffset() / sizeof(void*);
+
+        if (pFD->GetFieldType() != ELEMENT_TYPE_VALUETYPE)
+        {
+            if (pFD->IsObjRef())
+            {
+                if (gcPtrs[fieldStartIndex] == TYPE_GC_NONE)
+                {
+                    gcPtrs[fieldStartIndex] = TYPE_GC_REF;
+                    result++;
+                }
+            }
+        }
+        else
+        {
+            MethodTable * pFieldMT = pFD->GetApproxFieldTypeHandleThrowing().AsMethodTable();
+            if (pFieldMT->HasSameTypeDefAs(g_pByReferenceClass))
+            {
+                if (gcPtrs[fieldStartIndex] == TYPE_GC_NONE)
```

Comment:
If somebody creates union with overlapping byref and regular object ref, they are in trouble - it is very likely to work well. I think we should just fail with `COMPlusThrowHR(COR_E_BADIMAGEFORMAT);`: Add `if (gcPtrs[fieldStartIndex] != TYPE_GC_BYREF) COMPlusThrowHR(COR_E_BADIMAGEFORMAT);` as else block.